### PR TITLE
Add ClrObject.EnumerateReferencesWithFields

### PR DESCRIPTION
**Pull request by leculver**
Add a way to enumerate references of an object with a source of where that reference came from.

### CODE REVIEW

**Comment by leculver** [comment]
Need to test it (and add tests).  This is the rough direction I'm going in.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+        /// <summary>
+        /// The field this object was contained in.  This property may be null if metadata is not available, if
+        /// this reference came from a dependent handle/loader allocator, or if this came from an array.
+        /// Only valid to call if <see cref="IsObjectInterior"/> is true.
+        /// </summary>
+        public ClrInstanceField? Field { get; }
+
+        /// <summary>
+        /// If this reference came from a CollectableType, it means there's a collectable type which keeps this object alive.
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs:51]:
```suggestion
        /// If this reference came from a collectible type, it means there's a collectible type which keeps this object alive.
```

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                // Return the loader allocator handle if this doesn't have the high bits set.
+                if ((_offsetOrHandle & ValueMask) == _offsetOrHandle)
+                    return ValueMask & _offsetOrHandle;
+
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if this reference came from a LoaderAllocator handle (e.g. collectable types).
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs:67]:
```suggestion
        /// Returns true if this reference came from a LoaderAllocator handle (e.g. collectible types).
```

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+        /// <summary>
+        /// Returns true if this reference came from a dependent handle.
+        /// </summary>
+        public bool IsDepenendentHandle => (_offsetOrHandle & DependentFlag) == DependentFlag;
+
+        /// <summary>
+        /// Returns true if this reference came from a field in another object.
+        /// </summary>
+        public bool IsObjectInterior => LoaderAllocator == 0;
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs:79]: ❓

Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs:79]: Yeah, need to find a more descriptive term.  =)

Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs:79]: This is asserting that a dependent handle is also a field in another object. I don't understand why this is not:

```suggestion
        public bool IsObjectInterior => (_offsetOrHandle & OffsetFlag) == OffsetFlag;
```

Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrFieldReference.cs]: Ah right.  Also this isn't ready to merge, I was mostly only publishing it for a dev who was asking for the feature as a rough sketch of the design.

**Pull request merged**